static/elements/chromedash-approvals-dialog.js

```diff
@@ -1,42 +1,368 @@
+          margin: 0;
+        }
+
+        .comment_body {
+          background: var(--table-alternate-background);
+          padding: var(--content-padding-half);
+          white-space: pre-wrap;
+          width: 46em;
+          margin-bottom: var(--content-padding);
+        }
+
+        .controls {
+          padding: var(--content-padding);
+          text-align: right;
+        }
+
+        #show_all_checkbox {
+         float: left;
+        }
+
+        textarea {
+          padding: 4px;
+        }
       `];
   }
 
   openWithFeature(featureId) {
     this.featureId = featureId;
-    this.shadowRoot.querySelector('chromedash-dialog').open();
+    const p1 = window.csClient.getFeature(this.featureId).then(
+      (feature) => {
+        this.feature = feature;
+      });
+    const p2 = window.csClient.getApprovals(this.featureId).then(
+      (res) => {
+        this.approvals = res.approvals;
+        const numPending = this.approvals.filter((av) =>
+          ACTIVE_STATES.includes(av.state)).length;
+        this.subsetPending = (numPending > 0 &&
+                              numPending < APPROVAL_DEFS.length);
+        this.showAllIntents = numPending == 0;
+        this.changedApprovalsByField = new Map();
+        this.needsSave = false;
+      });
+    const p3 = window.csClient.getComments(this.featureId).then(
+      (res) => {
+        this.comments = res.comments;
+      });
+    Promise.all([p1, p2, p3]).then(() => {
+      // Clear out any previously typed comment.
+      this.shadowRoot.querySelector('#comment_area').value = '';
+      this.shadowRoot.querySelector('chromedash-dialog').open();
+    });
+  }
+
+
+  toggleShowAllIntents() {
+    this.showAllIntents = !this.showAllIntents;
+  }
+
+  formatDate(dateStr) {
+    return dateStr.split('.')[0]; // Ignore microseconds.
+  }
+
+  renderApprovalValue(approvalValue) {
+    const selectedValue = (
+      this.changedApprovalsByField.get(approvalValue.field_id) ||
+          approvalValue.state);
+    let placeholderOption = nothing;
+    if (approvalValue.state == -1) {
+      placeholderOption = html`
+        <option value="-1" selected>No value</option>
+      `;
+    }
+
+    return html`
+      <div class="approval_row">
+        <span class="set_by">${approvalValue.set_by}</span>
+        <span class="set_on">${this.formatDate(approvalValue.set_on)}</span>
+        <span class="appr_val">
+          ${approvalValue.set_by == this.signedInUser ? html`
+          <select
+            selected=${selectedValue}
+            data-field="${approvalValue.field_id}"
+            @change=${this.handleSelectChanged}
+          >
+              ${placeholderOption}
+              ${STATE_NAMES.map((valName) => html`
+                <option value="${valName[0]}"
+                  ?selected=${valName[0] == selectedValue}
+                 >${valName[1]}</option>`
+                )}
+            </select>` : html`
+            ${STATE_NAMES[approvalValue.state][1]}
```

Comment:
Could `STATE_NAMES[approvalValue.state]` be undefined?

Reply:
It should not be.  I have all the enum values for the state covered.